plugins/corsair/fu-corsair-device.c

```diff
@@ -0,0 +1,519 @@
+	gboolean ret;
+
+	ret = g_usb_device_interrupt_transfer(usb_device,
+					      self->epout,
+					      data,
+					      CORSAIR_CMD_SIZE,
+					      &actual_len,
+					      timeout,
+					      NULL,
+					      error);
+	if (!ret) {
+		g_prefix_error(error, "failed to write command: ");
+		return FALSE;
+	}
+	if (actual_len != CORSAIR_CMD_SIZE) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "wrong size written: %" G_GSIZE_FORMAT,
+			    actual_len);
+		return FALSE;
+	}
+
+	memset(data, 0, CORSAIR_CMD_SIZE);
+
+	ret = g_usb_device_interrupt_transfer(usb_device,
+					      self->epin,
+					      data,
+					      CORSAIR_CMD_SIZE,
+					      &actual_len,
+					      timeout,
+					      NULL,
+					      error);
+	if (!ret) {
+		g_prefix_error(error, "failed to get command response: ");
+		return FALSE;
+	}
+	if (actual_len != CORSAIR_CMD_SIZE) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "wrong size read: %" G_GSIZE_FORMAT,
+			    actual_len);
+		return FALSE;
+	}
+
+	if (data[2] != 0) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "device replied with error: %" G_GSIZE_FORMAT,
+			    data[2]);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_corsair_device_probe(FuDevice *device, GError **error)
+{
+	FuCorsairDevice *self = FU_CORSAIR_DEVICE(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	GUsbInterface *iface = NULL;
+	GUsbEndpoint *ep = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+	g_autoptr(GPtrArray) endpoints = NULL;
+
+	if (!FU_DEVICE_CLASS(fu_corsair_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	ifaces = g_usb_device_get_interfaces(usb_device, error);
+	if (ifaces == NULL || ifaces->len < 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "update interface not found");
+		return FALSE;
+	}
+
+	iface = g_ptr_array_index(ifaces, CORSAIR_VENDOR_INTERFACE_ID);
+	endpoints = g_usb_interface_get_endpoints(iface);
+	/* expecting to have two endpoints for communication */
+	if (endpoints == NULL || endpoints->len != 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "update interface endpoints not found");
+		return FALSE;
+	}
+
+	ep = g_ptr_array_index(endpoints, 0);
+	if (g_usb_endpoint_get_direction(ep) == G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST) {
+		self->epin = g_usb_endpoint_get_address(ep);
+		self->epout = g_usb_endpoint_get_address(g_ptr_array_index(endpoints, 1));
+	} else {
+		self->epin = g_usb_endpoint_get_address(g_ptr_array_index(endpoints, 0));
+		self->epout = g_usb_endpoint_get_address(ep);
+	}
+
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), CORSAIR_VENDOR_INTERFACE_ID);
+	return TRUE;
+}
+
+static gchar *
+fu_corsair_get_version(FuDevice *device, GError **error)
+{
+	guint32 version_raw;
+	guint8 data[CORSAIR_CMD_SIZE] = {0x08, 0x02, 0x13};
```

Comment:
can we replace any of these magic numbers with `#define`s?

Reply:
It was intentionally written with magics in order to not leak sensitive information. Can we leave it like this?

Reply:
I guess I'd prefer it to be documented so another Corsair employee can understand it (and extend it) in 5 years time, but if that's not possible it's fine.